src/scripts/sync_subclassof.py

```diff
@@ -0,0 +1,103 @@
+"""Create a robot template for purpose of syncing subclassof relations.
+
+Resources
+- GitHub issue: https://github.com/monarch-initiative/mondo-ingest/issues/92
+- GitHub PR: https://github.com/monarch-initiative/mondo-ingest/pull/363
+"""
+import os
+from argparse import ArgumentParser
+from datetime import datetime
+from pathlib import Path
+from typing import Dict, List, Tuple, Union
+
+import pandas as pd
+from oaklib import get_adapter
+from oaklib.interfaces.basic_ontology_interface import BasicOntologyInterface, RELATIONSHIP
+from oaklib.types import CURIE, URI
+
+HERE = Path(os.path.abspath(os.path.dirname(__file__)))
+SRC_DIR = HERE.parent
+ONTOLOGY_DIR = SRC_DIR / 'ontology'
+DEFAULT_OUTPATH = ONTOLOGY_DIR / 'reports' / 'sync-subclassof.robot.template.tsv'
+DEFAULT_ONTOLOGY_DB_PATH = ONTOLOGY_DIR / 'tmp' / 'mondo-ingest.db'
+ROBOT_ROW = [{
+    'term_id': 'ID',
+    'parent_id': 'SC %',
+    'source_ids': '> A oboInOwl:source SPLIT=|',
+}]
+
+
+def sync_subclassof(
+    outpath: str = DEFAULT_OUTPATH, ontology_db_path: str = DEFAULT_ONTOLOGY_DB_PATH
+):
+    """Run"""
+    print('Running - Synchronization pipeline: subClassOf')
+    oi: BasicOntologyInterface = get_adapter(ontology_db_path)
+
+    # Get all terms
+    print('- Collecting all terms')
+    t0 = datetime.now()
+    ids_all: List[Union[CURIE, URI]] = [x for x in oi.entities(filter_obsoletes=False)]
+    mondo_ids: List[CURIE] = [x for x in ids_all if x.startswith('MONDO:')]
+    t1 = datetime.now()
+    print(f'  Completed in {(t1 - t0).seconds} seconds')
+
+    # Get subclass edges
+    print('- Collecting and filtering relationships')
+    t0 = datetime.now()
+    rels_raw: List[RELATIONSHIP] = [x for x in oi.relationships(subjects=mondo_ids, predicates=['rdfs:subClassOf'])]
+    # - filter out non-mondo parents
+    rels_mondo_with_excluded = [x for x in rels_raw if x[2].startswith('MONDO:')]
+    # - filter excluded
+    exclusion_rels: List[RELATIONSHIP] = [x for x in oi.relationships(subjects=mondo_ids, predicates=['obo:mondo#excluded_subClassOf'])]
+    exclusion_rels_pairs: List[Tuple[CURIE, CURIE]] = [(x[0], x[2]) for x in exclusion_rels]
+    rels = [x for x in rels_mondo_with_excluded if (x[0], x[2]) not in exclusion_rels_pairs]
+    t1 = datetime.now()
+    print(f'  Completed in {(t1 - t0).seconds} seconds')
+
+    # Get subclass sources
+    print('- Collecting relationship sources')
+    print('  Estimated time: 1-2 hours')
+    t0 = datetime.now()
+    rels_and_sources: List[Tuple[RELATIONSHIP, List[CURIE]]] = []
+    for rel, metadatas in oi.relationships_metadata(rels):
```

Comment:
## Performance issue
So [this is now possible in OAK](https://github.com/INCATools/ontology-access-kit/pull/656). I'm implementing now, but it is a bit slow; I estimate it'll take 104 minutes to complete. I'm going to make an GH issue about that. If you want, I can change this to SPARQL, I think that might be faster.

_edit_: I created an [OAK PR](https://github.com/INCATools/ontology-access-kit/pull/659) that reduces this 104 minutes to just 30 seconds.
_edit_: ([Original comment](https://github.com/monarch-initiative/mondo-ingest/pull/363#discussion_r1336501888))

Reply:
Closing this thread now because no longer relevant. Don't need this method anymore for this PR.

Where I left off: I introduced 2 solutions, 1 of which was 10-100x faster or more for our use case. 
They weren't perfect, but they would fix a lot of use cases. However Chris didn't like them because they'd be a bad solution for some use cases.
Based on Chris' feedback, I introduced 2 more variations, 1 of which I could do in OAK but would be more difficult, and 1 which required an index addition to SemSql, which I don't have write permissions to. I was waiting for Chris' response, but he never got back.

Reply:
Thanks for keeping me posted on these issues, I am very interested in anything that can improve the general infrastructure.